blueprints/multi-interface-servers.md

```diff
@@ -0,0 +1,666 @@
+'service addresses'. And, of course, the property's values must be syntactically
+and semantically valid.
+
+Below are some examples of invalid `interfaces` property arrays, to which
+Traffic Ops MUST respond with a `400 Bad Request` and an appropriate Alert:
+
+Empty array
+```json
+{"interfaces": []}
+```
+Null property
+```json
+{"interfaces": null}
+```
+Missing property
+```json
+{}
+```
+Empty `ipAddresses` sub-property
+```json
+{
+	"interfaces": [
+		{
+			"ipAddresses": [],
+			"monitor": true,
+			"mtu": 1500,
+			"name": "eth0"
+		}
+	]
+}
+```
+Null `ipAddresses` sub-property
+```json
+{
+	"interfaces": [
+		{
+			"ipAddresses": null,
+			"monitor": true,
+			"mtu": 1500,
+			"name": "eth0"
+		}
+	]
+}
+```
+Missing `ipAddresses` sub-property
+```json
+{
+	"interfaces": [
+		{
+			"monitor": true,
+			"mtu": 1500,
+			"name": "eth0"
+		}
+	]
+}
+```
+No service addresses
+```json
+{
+	"interfaces": [
+		{
+			"ipAddresses": [
+				{
+					"address": "0.0.0.1/32",
+					"gateway": "10.0.0.1",
+					"serviceAddress": false
+				}
+			],
+			"monitor": true,
+			"mtu": 1500,
+			"name": "eth0"
+		}
+	]
+}
+```
+Too many interfaces with service addresses
+```json
+{
+	"interfaces": [
+		{
+			"ipAddresses": [
+				{
+					"address": "0.0.0.1/32",
+					"gateway": "10.0.0.1",
+					"serviceAddress": true
+				}
+			],
+			"monitor": true,
+			"mtu": 1500,
+			"name": "eth0"
+		},
+		{
+			"ipAddresses": [
+				{
+					"address": "f1d0::f00d/128",
+					"gateway": "dead::beef",
+					"serviceAddress": true
+				}
+			],
+			"monitor": true,
+			"mtu": 1500,
+			"name": "wnlp0"
+		}
+	]
+}
+```
+Too many service addresses of the same family/version
+```json
+{
+	"interfaces": [
+		{
+			"ipAddresses": [
+				{
+					"address": "0.0.0.1/32",
+					"gateway": "10.0.0.1",
+					"serviceAddress": true
+				},
+				{
+					"address": "0.0.0.2/32",
+					"gateway": "10.0.0.1",
+					"serviceAddress": true
+				}
+			],
+			"monitor": true,
+			"mtu": 1500,
+			"name": "eth0"
+		}
+	]
+}
+```
+##### `/servers/details`
+This endpoint is subject to the same changes described above for
+`/servers`/`/servers/{{ID}}` - though obviously the full output objects will
+differ.
+
+#### Client Impact
+The structures output by various clients will change, but no client code changes
+should be necessary.
+
+### Traffic Portal Impact
+The server edit/display/creation page will need to be updated to remove all of
+the properties of servers which no longer exist because they've been moved into
+the "interfaces" section, and a new set of interface-related input fields with
+controls for dynamically adding to them should be added.
+
+A sample screenshot is shown below.
+
+![](img/multi-interface-ui-sample.png "Mockup of proposed interface controls")
+
+If desired, the markup used for that sample may be found in
+[Appendix A](#appendix-a)
+
+Additionally, the properties of servers no longer present on a server object
```

Comment:
It might be beneficial however to show server.interface.ipAddress.address where serviceAddress=true however in the servers table for each server. I could see that being a property you might want to see when viewing a collection of servers. Something to think about.

Reply:
but you could have one or two of those - can you handle sorting by collection properties in TP tables?

Reply:
> but you could have one or two of those - can you handle sorting by collection properties in TP tables?

maybe?